src/slice.rs

```diff
@@ -629,3 +630,135 @@ macro_rules! s(
         &*&$crate::s![@parse ::std::marker::PhantomData::<$crate::Ix0>, [] $($t)*]
     };
 );
+
+/// Slicing information describing multiple mutable, disjoint slices.
+///
+/// It's unfortunate that we need `'a` and `A` to be parameters of the trait,
+/// but they're necessary until Rust supports generic associated types.
+pub trait MultiSlice<'a, A, D>
+where
+    A: 'a,
+    D: Dimension,
+{
+    /// The type of the slices created by `.multi_slice_move()`.
+    type Output;
+
+    /// Slice the raw view into multiple raw views, and dereference them.
+    ///
+    /// **Panics** if performing any individual slice panics or if the slices
+    /// are not disjoint (i.e. if they intersect).
+    fn multi_slice_move(&self, view: ArrayViewMut<'a, A, D>) -> Self::Output;
+}
+
+impl<'a, A, D> MultiSlice<'a, A, D> for ()
+where
+    A: 'a,
+    D: Dimension,
+{
+    type Output = ();
+
+    fn multi_slice_move(&self, _view: ArrayViewMut<'a, A, D>) -> Self::Output {}
+}
+
+impl<'a, A, D, Do0> MultiSlice<'a, A, D> for (SliceInfo<D::SliceArg, Do0>,)
```

Comment:
Do we need both the with-& and without impl? It seems like we don't need both for s![].

Reply:
We don't need both if the user only uses `s![]` to construct slicing information, because `s![]` always evaluates to a reference. However, it is possible to create owned `SliceInfo` instances directly, so supporting tuples of them is convenient for the user. If a user has a `(SliceInfo, SliceInfo)` instance, they could always call `.multi_slice_mut((&tuple.0, &tuple.1))`, but that's not as nice as `.multi_slice_mut(tuple)`. Or, if they have two `SliceInfo` instances, they could call `.multi_slice_mut((&a, &b))`, but that's not quite as nice as `.multi_slice_mut((a, b))`.

I wouldn't be strongly opposed to removing support for multi-slicing with `(SliceInfo, SliceInfo)`, because I doubt it would be used often. However, there isn't much cost to keeping it in the PR another than a little more code and maybe slightly longer compile time.

Reply:
Fwiw, in the future, I'd like to change the parameter in all the slicing methods from `&SliceInfo<D::SliceArg, Do>` to `T: CanSlice<D>`. This simplifies handling of const versus dyn dimension types, avoids the need for extra type parameters on slicing methods when we add support for ellipses in slicing, and allows us to be generic over whether the slicing info is owned (`SliceInfo`) or a reference (`&SliceInfo`). This last advantage would also avoid the need for separate `MultiSlice` implementations for tuples of `SliceInfo` and tuples of `&SliceInfo`. (We'd implement `MultiSlice` for `(T1, T2, ...)` where `T1: CanSlice<D>, T2: CanSlice<D>, ...`.)

Reply:
I lean towards you ain't gonna need it and having 95% of the functionality for just the one impl. Yes, I fear I might lean towards a foolish stinginess sometimes. (So I need guidance on how to productively make the crate both simple to understand and fast to compile.)

Reply:
Okay, I've updated the PR.